music21/expressions.py

```diff
@@ -625,99 +773,177 @@ def name(self) -> str:
-        elif pitch.isValidAccidentalName(newAccid):
-            self._accid = pitch.makeAccidentalName(newAccid)
-
-    @property
-    def size(self) -> interval.IntervalBase | None:
-        if self._size is None:
-            raise ExpressionException(
-                'Cannot compute mordent size without srcPitch unless .size has been '
-                + 'explicitly set (use mordent.getSize(srcPitch) instead.)'
-            )
-        return self._size
+    def accidental(self) -> pitch.Accidental | None:
+        '''
+        This is the GeneralMordent's accidental.
+        '''
+        return self._accidental
 
-    @size.setter
-    def size(self, newSize: interval.IntervalBase | None):
-        if self._accid is not None:
-            raise ExpressionException('Cannot set mordent size if accid already set.')
-        self._size = newSize
+    @accidental.setter
+    def accidental(self, newAccidental: pitch.Accidental | None):
+        self._accidental = newAccidental
 
     @property
     def direction(self) -> str:
         return self._direction
 
     def getSize(
         self,
-        srcPitch: pitch.Pitch,
+        srcObj: note.GeneralNote,
+        *,
+        keySig: key.KeySignature | None = None,
     ) -> interval.IntervalBase:
-        if self._size is not None:
-            return self._size
-
-        if self.direction not in ('up', 'down'):
+        '''
+        Returns the size of the mordent's interval, given a source note and
+        an optional key signature.  If the key signature is not specified, the
+        source note's context is searched for the current key signature, and if
+        there is no such key signature, a key signature with no sharps and no flats
+        will be used.  Any `accidental` that has been set on the mordent will also
+        be taken into account.  If no `accidental` has been set, the appropriate
+        accidental from the key signature will be used.
+
+        If keySig is specified, this can be considered to be a theoretical question:
+        "If this particular mordent were to be attached to this note, in this key,
+        what would the size of the mordent interval be?"
+        '''
+        if self._direction not in ('up', 'down'):
             raise ExpressionException('Cannot compute mordent size if I do not know its direction')
 
-        otherPitch: pitch.Pitch = copy.deepcopy(srcPitch)
-        otherPitch.accidental = None
+        if not srcObj.pitches:
+            # perfect unison
+            return interval.Interval('P1')
+
+        if keySig is None:
+            keySig = srcObj.getContextByClass(key.KeySignature)
+            if keySig is None:
+                keySig = key.KeySignature(0)
+
+        srcPitch: pitch.Pitch = srcObj.pitches[-1]
+
+        ornamentalPitch: pitch.Pitch = copy.deepcopy(srcPitch)
+        ornamentalPitch.accidental = None
+        if ornamentalPitch.octave is None:
+            ornamentalPitch.octave = ornamentalPitch.implicitOctave
```

Comment:
Ha!  I wish I had remembered it.  It caused a test failure where a half-step up from B (implicit octave = 4) was C (implicit octave = 4), so that half step up turned into a major 7th down.  Took me several minutes to figure out why.